cli/task_runner.rs

```diff
@@ -0,0 +1,506 @@
+      && !context.args.iter().any(|s| s.starts_with('-'))
+    {
+      // run with deno task instead
+      let mut args = Vec::with_capacity(context.args.len());
+      args.push("task".to_string());
+      args.extend(context.args.iter().skip(1).cloned());
+
+      let mut state = context.state;
+      state.apply_env_var(USE_PKG_JSON_HIDDEN_ENV_VAR_NAME, "1");
+      return ExecutableCommand::new(
+        "deno".to_string(),
+        std::env::current_exe().unwrap(),
+      )
+      .execute(ShellCommandContext {
+        args,
+        state,
+        ..context
+      });
+    }
+
+    // fallback to running the real npm command
+    let npm_path = match context.state.resolve_command_path("npm") {
+      Ok(path) => path,
+      Err(err) => {
+        let _ = context.stderr.write_line(&format!("{}", err));
+        return Box::pin(futures::future::ready(
+          ExecuteResult::from_exit_code(err.exit_code()),
+        ));
+      }
+    };
+    ExecutableCommand::new("npm".to_string(), npm_path).execute(context)
+  }
+}
+
+pub struct NodeCommand;
+
+impl ShellCommand for NodeCommand {
+  fn execute(
+    &self,
+    context: ShellCommandContext,
+  ) -> LocalBoxFuture<'static, ExecuteResult> {
+    // run with deno if it's a simple invocation, fall back to node
+    // if there are extra flags
+    let mut args = Vec::with_capacity(context.args.len());
+    if context.args.len() > 1
+      && (
+        context.args[0].starts_with('-') // has a flag
+        || !matches!(
+          MediaType::from_str(&context.args[0]),
+          MediaType::Cjs | MediaType::Mjs | MediaType::JavaScript
+        )
+        // not a script file
+      )
+    {
+      return ExecutableCommand::new(
+        "node".to_string(),
+        "node".to_string().into(),
+      )
+      .execute(context);
+    }
+    args.extend(["run", "-A"].into_iter().map(|s| s.to_string()));
+    args.extend(context.args.iter().cloned());
+
+    let mut state = context.state;
+    state.apply_env_var(USE_PKG_JSON_HIDDEN_ENV_VAR_NAME, "1");
+    ExecutableCommand::new("deno".to_string(), std::env::current_exe().unwrap())
+      .execute(ShellCommandContext {
+        args,
+        state,
+        ..context
+      })
+  }
+}
+
+pub struct NodeGypCommand;
+
+impl ShellCommand for NodeGypCommand {
+  fn execute(
+    &self,
+    context: ShellCommandContext,
+  ) -> LocalBoxFuture<'static, ExecuteResult> {
+    // at the moment this shell command is just to give a warning if node-gyp is not found
+    // in the future, we could try to run/install node-gyp for the user with deno
+    if which::which("node-gyp").is_err() {
+      log::warn!("{}: node-gyp was used in a script, but was not listed as a dependency. Either add it as a dependency or install it globally (e.g. `npm install -g node-gyp`)", crate::colors::yellow("warning"));
+    }
+    ExecutableCommand::new(
+      "node-gyp".to_string(),
+      "node-gyp".to_string().into(),
+    )
+    .execute(context)
+  }
+}
+
```

Comment:
These were added in this PR (the rest was pretty much a straight move, w some minor changes).

Reply:
Why does it use -A ? It would be better if it prompts the user for permissions

Reply:
It uses -A because it's executing binaries like `vite` that don't have a concept of permission flags.

Reply:
Can you elaborate more, isn't all JavaScript code permission unaware (the concept exists at deno runtime) what's special about vite in that case?

Reply:
Say you have this in a package.json:

```json
{
  "scripts": {
    "dev": "vite dev"
  }
}
```

The `vite` command doesn't have the concept of permission flags.